docs/code-organisation-guidelines.md

```diff
@@ -0,0 +1,65 @@
+  within `util` would have a name prefix `Pos.Util`, modules within `networking` might have a prefix
+  of `Pos.Network` and so on. This means that when looking at the imports for a module, it is
+  immediately obvious which package each import comes from. As an example of a file that needs to be
+  renamed/moved the file `core/Pos/Aeson/Core.hs` should be renamed to `core/Pos/Core/Aeson.hs` so
+  that if it is imported in another module its name `Pos.Core.Aeson` would immediately suggest its
+  location in the source tree.
+
+* All library code for a Haskell packages should be in the `Pos` directory of that package.
+  Currently some packages use `src` and then `src` has both `Pos` and `Test` sub-directories. Since
+  all test code should be in the `test` sub directory (see below), we should choose `Pos` over
+  `src/Pos`. For example `lib` currently has testing related code in the `src` directory and if
+  that test code moves to the `test` directory then there is no need for `src`.
+
+* Tests for a Haskell package should always be a part of the same package as the implementation of
+  the code it is testing. Test code should not be included in the library that a package provides.
+  All test code for a Haskell package should be in the `test` directory of that package and modules
+  with testing related code should have module names staring with `Test` and then reflecting the
+  Haskell package which the code is testing, so test code for the `core` package would have all its
+  test module names begin with `Test.Pos.Core`. An example of where the current code base
+  contravenes these guidelines is that `core` and `txp` currently don't define any tests, but at
+  least some of the functionality in these two modules are tested in `lib` package.
+
+* Arbitrary instances for testing should never be in the library source tree and should not be
+  compiled into the library defined by the package. Instead, Arbitrary instances should be under the
+  `test` directory as per the previous point. When Arbitrary instances or other testing helper
+  functionality needs to be used by other packages, a test specific package should be created. For
+  instance, the `core` package may have `Arbitrary` instances that are need for tests in the `txp`
+  package. In this case, there will be a cabal file `core/test/cardano-sl-core-test.cabal` which
+  exports the required modules, all of which would have a `Test.Pos.Core` name prefix.
+  [PR #2588](https://github.com/input-output-hk/cardano-sl/pull/2588) is an example of doing this
+  for the `crypto` package.
+
+* The library that a Haskell package provides should never include or export any code that is
+  not part of the API the library provides to clients of the library. It is however valid for
+  a library to provide interface functions to its internals that are required for testing or
+  benchmarking, thus becoming part of the library's API.
```

Comment:
The pattern I like is: expose a `Foo.Internal` module that exports *everything* with a big disclaimer that it's not public API and might change, invariants might break, etc. and expose the "public/safe" API from `Foo`. Should we create a convention like this how library internals are exposed?

Reply:
I think this is possibly a second or third step.

Reply:
I think step 0 might be to put all the internals at the end of the exports list, "guarded" by a Haddock annotation "internals: testing only" kind of thing 😀